core/src/cas/eav.rs

```diff
@@ -0,0 +1,173 @@
+/// provides a simple and flexible interface to define relationships between AddressableContent
+pub trait EntityAttributeValueStorage {
+    /// adds the given EntityAttributeValue to the EntityAttributeValueStorage
+    /// append only storage
+    /// eavs are retrieved through constraint based lookups
+    /// @see fetch_eav
+    fn add_eav(&mut self, eav: &EntityAttributeValue) -> Result<(), HolochainError>;
+    /// fetches the set of EntityAttributeValues that match constraints
+    /// None = no constraint
+    /// Some(Entity) = requires the given entity (e.g. all a/v pairs for the entity)
+    /// Some(Attribute) = requires the given attribute (e.g. all links)
+    /// Some(Value) = requires the given value (e.g. all entities referencing an Address)
+    fn fetch_eav(&self, entity: Option<Entity>, attribute: Option<Attribute>, value: Option<Value>) -> Result<HashSet<EntityAttributeValue>, HolochainError>;
+}
+
+#[cfg(test)]
+pub mod tests {
+    use error::HolochainError;
+    use cas::eav::Entity;
+    use cas::eav::Attribute;
+    use cas::eav::Value;
+    use cas::eav::EntityAttributeValueStorage;
+    use cas::eav::EntityAttributeValue;
+    use std::collections::HashSet;
+    use cas::content::Content;
+    use cas::content::AddressableContent;
+    use cas::content::tests::ExampleAddressableContent;
+
+    pub struct ExampleEntityAttributeValueStorage {
+        eavs: HashSet<EntityAttributeValue>,
+    }
+
+    impl ExampleEntityAttributeValueStorage {
+        pub fn new() -> ExampleEntityAttributeValueStorage {
+            ExampleEntityAttributeValueStorage{
+                eavs: HashSet::new(),
+            }
+        }
+    }
+
+    impl EntityAttributeValueStorage for ExampleEntityAttributeValueStorage {
+        fn add_eav(&mut self, eav: &EntityAttributeValue) -> Result<(), HolochainError> {
+            self.eavs.insert(eav.clone());
+            Ok(())
+        }
+
+        fn fetch_eav(&self, entity: Option<Entity>, attribute: Option<Attribute>, value: Option<Value>) -> Result<HashSet<EntityAttributeValue>, HolochainError> {
+            let filtered = self.eavs.iter().cloned()
+                    .filter(|eav| {
+                        match entity {
+                            Some(ref e) => &eav.entity() == e,
+                            None => true,
+                        }
+                    })
+                    .filter(|eav| {
+                        match attribute {
+                            Some(ref a) => &eav.attribute() == a,
+                            None => true,
+                        }
+                    })
+                    .filter(|eav| {
+                        match value {
+                            Some(ref v) => &eav.value() == v,
+                            None => true,
+                        }
+                    })
+                    .collect::<HashSet<EntityAttributeValue>>();
+            Ok(filtered)
+        }
```

Comment:
Cool, I like this! We might still add some more helper public methods with less arguments that internally calls this. 

For example:
 ```Rust
fn fetch_relations_from(&self, entity: &Entity) -> Result<HashSet<EntityAttributeValue>, HolochainError> {
  self.fetch_eav(Some(entity), None, None)
}
```

Reply:
@ddd-mtl oh yeah for sure, i just thought i'd get the low level constraint logic in place for this PR

IIRC combined CAS + EAV storage affords entire query languages based on constraints

e.g. https://en.wikipedia.org/wiki/Datalog

we don't need anything near that for the developer preview though ;)